File: app/src/main/java/de/westnordost/streetcomplete/quests/opening_hours/AddOpeningHoursForm.java
package de.westnordost.streetcomplete.quests.opening_hours;

import android.os.Bundle;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.PopupMenu;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;

import java.util.ArrayList;

import javax.inject.Inject;

import de.westnordost.streetcomplete.Injector;
import de.westnordost.streetcomplete.R;
import de.westnordost.streetcomplete.quests.AbstractQuestFormAnswerFragment;
import de.westnordost.streetcomplete.quests.opening_hours.adapter.AddOpeningHoursAdapter;
import de.westnordost.streetcomplete.quests.opening_hours.adapter.OpeningMonthsRow;
import de.westnordost.streetcomplete.util.AdapterDataChangedWatcher;
import de.westnordost.streetcomplete.util.Serializer;


import static android.view.Menu.NONE;

public class AddOpeningHoursForm extends AbstractQuestFormAnswerFragment
{
	public static final String OPENING_HOURS = "opening_hours";
	public static final String NO_SIGN = "no_sign";

	private static final String	OPENING_HOURS_DATA = "oh_data",
								IS_ADD_MONTHS_MODE = "oh_add_months";

	private boolean isAlsoAddingMonths;
	private AddOpeningHoursAdapter openingHoursAdapter;

	@Inject Serializer serializer;

	@Override public View onCreateView(LayoutInflater inflater, ViewGroup container,
							 Bundle savedInstanceState)
	{
		View view = super.onCreateView(inflater, container, savedInstanceState);

		Injector.instance.getApplicationComponent().inject(this);

		addOtherAnswers();

		View contentView = setContentView(R.layout.quest_opening_hours);

		initOpeningHoursAdapter(contentView, savedInstanceState);

		Button addTimes = contentView.findViewById(R.id.btn_add);
		addTimes.setOnClickListener(this::onClickAddButton);

		return view;
	}

	private void initOpeningHoursAdapter(View contentView, Bundle savedInstanceState)
	{
		ArrayList<OpeningMonthsRow> viewData;
		if(savedInstanceState != null)
		{
			viewData = serializer.toObject(savedInstanceState.getByteArray(OPENING_HOURS_DATA),ArrayList.class);
			isAlsoAddingMonths = savedInstanceState.getBoolean(IS_ADD_MONTHS_MODE);
		}
		else
		{
			viewData = new ArrayList<>();
			viewData.add(new OpeningMonthsRow());
			isAlsoAddingMonths = false;
		}

		openingHoursAdapter = new AddOpeningHoursAdapter(viewData, getActivity(), getCountryInfo());
		openingHoursAdapter.setDisplayMonths(isAlsoAddingMonths);
		openingHoursAdapter.registerAdapterDataObserver(new AdapterDataChangedWatcher(this::checkIsFormComplete));
		RecyclerView openingHoursList = contentView.findViewById(R.id.opening_hours_list);
		openingHoursList.setLayoutManager(new LinearLayoutManager(getActivity(), LinearLayoutManager.VERTICAL, false));
		openingHoursList.setAdapter(openingHoursAdapter);
		openingHoursList.setNestedScrollingEnabled(false);
		checkIsFormComplete();
	}

	private void addOtherAnswers()
	{
		addOtherAnswer(R.string.quest_openingHours_no_sign, this::confirmNoSign);
		addOtherAnswer(R.string.quest_openingHours_answer_no_regular_opening_hours, this::showInputCommentDialog);
		addOtherAnswer(R.string.quest_openingHours_answer_247, this::showConfirm24_7Dialog);
		addOtherAnswer(R.string.quest_openingHours_answer_seasonal_opening_hours, this::changeToMonthsMode);
	}

	private void onClickAddButton(View v)
	{
		if(!isAlsoAddingMonths)
		{
			openingHoursAdapter.addNewWeekdays();
		}
		else
		{
			PopupMenu m = new PopupMenu(getActivity(), v);
			m.getMenu().add(NONE,0,NONE,R.string.quest_openingHours_add_weekdays);
			m.getMenu().add(NONE,1,NONE,R.string.quest_openingHours_add_months);
			m.setOnMenuItemClickListener(item ->
			{
				if(0 == item.getItemId()) openingHoursAdapter.addNewWeekdays();
				else if(1 == item.getItemId()) openingHoursAdapter.addNewMonths();
				return true;
			});
			m.show();
		}
	}

	@Override public void onSaveInstanceState(Bundle outState)
	{
		super.onSaveInstanceState(outState);
		outState.putByteArray(OPENING_HOURS_DATA, serializer.toBytes(openingHoursAdapter.getViewData()));
		outState.putBoolean(IS_ADD_MONTHS_MODE, isAlsoAddingMonths);
	}

	@Override protected void onClickOk()
	{
		Bundle answer = new Bundle();
		answer.putString(OPENING_HOURS, getOpeningHoursString());
		applyAnswer(answer);
	}

	private void showInputCommentDialog()
	{
		View view = LayoutInflater.from(getActivity()).inflate(R.layout.quest_opening_hours_comment, null);
		final EditText editText = view.findViewById(R.id.commentInput);

		new AlertDialog.Builder(getContext())
				.setTitle(R.string.quest_openingHours_comment_title)
				.setView(view)
				.setPositiveButton(android.R.string.ok, (dialog, which) ->
				{
					String txt = editText.getText().toString().replaceAll("\"","").trim();
					if(txt.isEmpty())
					{
						new AlertDialog.Builder(getContext())
								.setMessage(R.string.quest_openingHours_emptyAnswer)
								.setPositiveButton(R.string.ok, null)
								.show();
						return;
					}

					Bundle answer = new Bundle();
					answer.putString(OPENING_HOURS, "\""+txt+"\"");
					applyAnswer(answer);
				})
				.setNegativeButton(android.R.string.cancel, null)
				.show();
	}

	private void showConfirm24_7Dialog()
	{
		new AlertDialog.Builder(getActivity())
				.setMessage(R.string.quest_openingHours_24_7_confirmation)
				.setPositiveButton(android.R.string.yes, (dialog, which) ->
				{
					Bundle answer = new Bundle();
					answer.putString(OPENING_HOURS, "24/7");
					applyAnswer(answer);
				})
				.setNegativeButton(android.R.string.no, null)
				.show();
	}

	private void confirmNoSign()
	{
		new AlertDialog.Builder(getActivity())
			.setTitle(R.string.quest_generic_confirmation_title)
			.setPositiveButton(R.string.quest_generic_confirmation_yes, (dialog, which) ->
			{
				Bundle data = new Bundle();
				data.putBoolean(NO_SIGN, true);
				applyAnswer(data);
			})
			.setNegativeButton(R.string.quest_generic_confirmation_no, null)
			.show();
	}

	private void changeToMonthsMode()
	{
		isAlsoAddingMonths = true;
		openingHoursAdapter.changeToMonthsMode();
	}

	@Override public boolean isFormComplete() { return !getOpeningHoursString().isEmpty(); }

	private String getOpeningHoursString()
	{
		return TextUtils.join(";", openingHoursAdapter.createData());
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/quests/opening_hours/AddOpeningHoursForm.kt
package de.westnordost.streetcomplete.quests.opening_hours

import android.os.Bundle
import android.view.Menu.NONE
import android.view.View
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.widget.PopupMenu
import androidx.core.view.isGone
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import de.westnordost.osm_opening_hours.parser.toOpeningHours
import de.westnordost.osm_opening_hours.parser.toOpeningHoursOrNull
import de.westnordost.streetcomplete.R
import de.westnordost.streetcomplete.databinding.QuestOpeningHoursBinding
import de.westnordost.streetcomplete.databinding.QuestOpeningHoursCommentBinding
import de.westnordost.streetcomplete.osm.opening_hours.parser.toOpeningHoursRows
import de.westnordost.streetcomplete.quests.AbstractOsmQuestForm
import de.westnordost.streetcomplete.quests.AnswerItem
import de.westnordost.streetcomplete.quests.opening_hours.adapter.OpeningHoursAdapter
import de.westnordost.streetcomplete.quests.opening_hours.adapter.OpeningMonthsRow
import de.westnordost.streetcomplete.quests.opening_hours.adapter.OpeningWeekdaysRow
import de.westnordost.streetcomplete.view.AdapterDataChangedWatcher
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

class AddOpeningHoursForm : AbstractOsmQuestForm<OpeningHoursAnswer>() {

    override val contentLayoutResId = R.layout.quest_opening_hours
    private val binding by contentViewBinding(QuestOpeningHoursBinding::bind)

    override val buttonPanelAnswers get() =
        if (isDisplayingPreviousOpeningHours) {
            listOf(
                AnswerItem(R.string.quest_generic_hasFeature_no) { setAsResurvey(false) },
                AnswerItem(R.string.quest_generic_hasFeature_yes) {
                    applyAnswer(RegularOpeningHours(
                        element.tags["opening_hours"]!!.toOpeningHours(lenient = true)
                    ))
                }
            )
        } else {
            emptyList()
        }

    override val otherAnswers = listOf(
        AnswerItem(R.string.quest_openingHours_no_sign) { confirmNoSign() },
        AnswerItem(R.string.quest_openingHours_answer_no_regular_opening_hours) { showInputCommentDialog() },
        AnswerItem(R.string.quest_openingHours_answer_247) { showConfirm24_7Dialog() },
        AnswerItem(R.string.quest_openingHours_answer_seasonal_opening_hours) {
            setAsResurvey(false)
            openingHoursAdapter.changeToMonthsMode()
        }
    )

    private lateinit var openingHoursAdapter: OpeningHoursAdapter

    private var isDisplayingPreviousOpeningHours: Boolean = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        openingHoursAdapter = OpeningHoursAdapter(requireContext())
        openingHoursAdapter.firstDayOfWorkweek = countryInfo.firstDayOfWorkweek
        openingHoursAdapter.regularShoppingDays = countryInfo.regularShoppingDays
        openingHoursAdapter.locale = countryInfo.userPreferredLocale
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        if (savedInstanceState != null) {
            onLoadInstanceState(savedInstanceState)
        } else {
            initStateFromTags()
        }

        openingHoursAdapter.registerAdapterDataObserver(AdapterDataChangedWatcher { checkIsFormComplete() })

        binding.openingHoursList.layoutManager = LinearLayoutManager(activity, RecyclerView.VERTICAL, false)
        binding.openingHoursList.adapter = openingHoursAdapter
        binding.openingHoursList.isNestedScrollingEnabled = false
        checkIsFormComplete()

        binding.addTimesButton.setOnClickListener { onClickAddButton(it) }
    }

    private fun onClickAddButton(v: View) {
        val rows = openingHoursAdapter.rows

        val addMonthAvailable = rows.any { it is OpeningMonthsRow }
        val addTimeAvailable = rows.isNotEmpty() && rows.last() is OpeningWeekdaysRow
        val addOffDayAvailable = rows.isNotEmpty() && rows.last() is OpeningWeekdaysRow

        if (addMonthAvailable || addTimeAvailable || addOffDayAvailable) {
            val popup = PopupMenu(requireContext(), v)
            if (addTimeAvailable) popup.menu.add(NONE, 0, NONE, R.string.quest_openingHours_add_hours)
            popup.menu.add(NONE, 1, NONE, R.string.quest_openingHours_add_weekdays)
            if (addOffDayAvailable) popup.menu.add(NONE, 2, NONE, R.string.quest_openingHours_add_off_days)
            if (addMonthAvailable) popup.menu.add(NONE, 3, NONE, R.string.quest_openingHours_add_months)
            popup.setOnMenuItemClickListener { item ->
                when (item.itemId) {
                    0 -> openingHoursAdapter.addNewHours()
                    1 -> openingHoursAdapter.addNewWeekdays()
                    2 -> openingHoursAdapter.addNewOffDays()
                    3 -> openingHoursAdapter.addNewMonths()
                }
                true
            }
            popup.show()
        } else {
            openingHoursAdapter.addNewWeekdays()
        }
    }

    private fun initStateFromTags() {
        val oh = element.tags["opening_hours"]
        val rows = oh?.toOpeningHoursOrNull(lenient = true)?.toOpeningHoursRows()
        if (rows != null) {
            openingHoursAdapter.rows = rows.toMutableList()
            setAsResurvey(true)
        } else {
            setAsResurvey(false)
        }
    }

    private fun onLoadInstanceState(savedInstanceState: Bundle) {
        openingHoursAdapter.rows = Json.decodeFromString(savedInstanceState.getString(OPENING_HOURS_DATA)!!)
        setAsResurvey(savedInstanceState.getBoolean(IS_DISPLAYING_PREVIOUS_HOURS))
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)

        outState.putString(OPENING_HOURS_DATA, Json.encodeToString(openingHoursAdapter.rows))
        outState.putBoolean(IS_DISPLAYING_PREVIOUS_HOURS, isDisplayingPreviousOpeningHours)
    }

    override fun onClickOk() {
        applyAnswer(RegularOpeningHours(openingHoursAdapter.createOpeningHours()))
    }

    private fun showInputCommentDialog() {
        val dialogBinding = QuestOpeningHoursCommentBinding.inflate(layoutInflater)

        AlertDialog.Builder(requireContext())
            .setTitle(R.string.quest_openingHours_comment_title)
            .setView(dialogBinding.root)
            .setPositiveButton(android.R.string.ok) { _, _ ->
                val txt = dialogBinding.commentInput.text.toString().replace("\"", "").trim()
                if (txt.isEmpty()) {
                    AlertDialog.Builder(requireContext())
                        .setMessage(R.string.quest_openingHours_emptyAnswer)
                        .setPositiveButton(android.R.string.ok, null)
                        .show()
                } else {
                    applyAnswer(DescribeOpeningHours(txt))
                }
            }
            .setNegativeButton(android.R.string.cancel, null)
            .show()
    }

    private fun setAsResurvey(resurvey: Boolean) {
        openingHoursAdapter.isEnabled = !resurvey
        isDisplayingPreviousOpeningHours = resurvey
        binding.addTimesButton.isGone = resurvey
        updateButtonPanel()
    }

    private fun showConfirm24_7Dialog() {
        AlertDialog.Builder(requireContext())
            .setMessage(R.string.quest_openingHours_24_7_confirmation)
            .setPositiveButton(R.string.quest_generic_hasFeature_yes) { _, _ ->
                applyAnswer(AlwaysOpen)
            }
            .setNegativeButton(R.string.quest_generic_hasFeature_no, null)
            .show()
    }

    private fun confirmNoSign() {
        AlertDialog.Builder(requireContext())
            .setTitle(R.string.quest_generic_confirmation_title)
            .setPositiveButton(R.string.quest_generic_confirmation_yes) { _, _ -> applyAnswer(NoOpeningHoursSign) }
            .setNegativeButton(R.string.quest_generic_confirmation_no, null)
            .show()
    }

    override fun isFormComplete() = openingHoursAdapter.rows.isNotEmpty() && !isDisplayingPreviousOpeningHours

    companion object {
        private const val OPENING_HOURS_DATA = "oh_data"
        private const val IS_DISPLAYING_PREVIOUS_HOURS = "oh_is_displaying_previous_hours"
    }
}
